pkg/storage/etcd3/watcher.go

```diff
@@ -322,7 +322,13 @@ func prepareObjs(ctx context.Context, e *event, client *clientv3.Client, codec r
 		if err != nil {
 			return nil, nil, err
 		}
-		oldObj, err = decodeObj(codec, versioner, getResp.Kvs[0].Value, getResp.Kvs[0].ModRevision)
+		rev := getResp.Kvs[0].ModRevision
+		if e.isDeleted {
+			// Note that this sends the *old* object with the etcd revision for the time at
+			// which it gets deleted. Users use that revision to tell ordering.
+			rev = e.rev
+		}
+		oldObj, err = decodeObj(codec, versioner, getResp.Kvs[0].Value, rev)
```

Comment:
The etcd2 code is written in first comment too:

https://github.com/kubernetes/kubernetes/blob/995f022808d1407897a5b23373484ac683534390/pkg/storage/etcd/etcd_watcher.go#L437-L442

[watch.Event](https://github.com/kubernetes/kubernetes/blob/3111985564f4bfc769111c5708a01d9d711edf70/pkg/watch/watch.go#L48-L57) contains only one object.

> So the rev of the oldObj should not matter at all for ordering purpose. No?

I have the same concern. The object is actually returned via Informer too.